File: brokerage-apis/examples/node/accounts-examples.mjs
/**
 * Accounts API examples: list, balances, positions, order status, transactions.
 *
 *     node accounts-examples.mjs
 */

import { ApiError, BrokerageClient, Dec } from './brokerage-client.mjs';

const pad = (s, n) => String(s).padEnd(n);
const rpad = (s, n) => String(s).padStart(n);

// ---------------------------------------------------------------------------
// GET /accounts
// ---------------------------------------------------------------------------

/**
 * Every account reachable with the current grant.
 *
 * `accountId` is the opaque handle used everywhere else. `accountNumber` is
 * the customer-facing value and arrives masked unless the grant carries
 * `accounts:read:full` -- do not key anything on it.
 */
export async function listAccounts(client) {
  const { items: accounts } = await client.get('/accounts');

  for (const account of accounts) {
    console.log(
      `${account.accountId}  ${rpad(account.accountNumber, 10)}  ` +
        `${pad(account.type, 7)} ${pad(account.registration, 16)} ` +
        `${pad(account.status, 10)} options=${account.optionsLevel}`,
    );

    // Affordances are state-dependent: a RESTRICTED account carries no
    // place-order link, so a UI can disable the trade button without knowing
    // anything about compliance rules.
    if (!client.can(account, 'place-order')) {
      console.log('    (not trade-enabled with this grant)');
    }
  }

  return accounts;
}

// ---------------------------------------------------------------------------
// GET /accounts/{accountId}/balances
// ---------------------------------------------------------------------------

/**
 * Cash, buying power, and margin. Reached by following the `balances` link
 * rather than building the URL, so the client survives a path change.
 */
export async function showBalances(client, account) {
  const balances = await client.follow(account, 'balances');

  console.log(`\nBalances as of ${balances.asOf} (${balances.currency})`);
  console.log(`  Net liquidation : ${rpad(Dec.format(balances.netLiquidation), 14)}`);
  console.log(`  Cash (settled)  : ${rpad(Dec.format(balances.cash.settled), 14)}`);
  console.log(`  Cash (total)    : ${rpad(Dec.format(balances.cash.total), 14)}`);
  console.log(`  Withdrawable    : ${rpad(Dec.format(balances.cash.withdrawable), 14)}`);

  // Stock and option buying power differ materially in a margin account --
  // sizing an option order off stock buying power gets it rejected.
  console.log('  Buying power');
  console.log(`    stock         : ${rpad(Dec.format(balances.buyingPower.stock), 14)}`);
  console.log(`    option        : ${rpad(Dec.format(balances.buyingPower.option), 14)}`);
  if (balances.buyingPower.dayTrading) {
    console.log(`    day trading   : ${rpad(Dec.format(balances.buyingPower.dayTrading), 14)}`);
  }

  if (balances.margin) {
    console.log(`  Excess liquidity: ${rpad(Dec.format(balances.margin.excessLiquidity), 14)}`);
    const call = balances.margin.marginCall;
    if (call) {
      // A margin call is the one balance field worth alerting on.
      console.log(`  ** ${call.type}: ${Dec.format(call.amount)} due ${call.dueDate} **`);
    }
  }

  if (balances.patternDayTrader) {
    console.log(`  Day trades left : ${balances.dayTradesRemaining}`);
  }

  return balances;
}

// ---------------------------------------------------------------------------
// GET /accounts/{accountId}/positions
// ---------------------------------------------------------------------------

export async function showPositions(client, account) {
  const positions = await client.collect(client.link(account, 'positions'));

  if (positions.length === 0) {
    console.log('\nNo open positions.');
    return positions;
  }

  console.log(
    `\n${pad('Symbol', 24)} ${pad('Side', 6)} ${rpad('Qty', 8)} ` +
      `${rpad('Avg cost', 11)} ${rpad('Market value', 14)} ${rpad('Unrealized', 13)}`,
  );
  console.log('-'.repeat(80));

  let totalUnrealized = '0';
  for (const position of positions) {
    const { instrument } = position;

    // Option positions carry the 21-character OSI symbol, which is unreadable
    // in a list. Build something human-facing from the option detail instead.
    const label =
      instrument.type === 'OPTION'
        ? `${instrument.option.underlyingSymbol} ${instrument.option.expiration.slice(2)} ` +
          `${Dec.format(instrument.option.strike, 0)}${instrument.option.optionType[0]}`
        : instrument.symbol;

    totalUnrealized = Dec.add(totalUnrealized, position.unrealizedPnl);

    console.log(
      `${pad(label, 24)} ${pad(position.side, 6)} ` +
        `${rpad(Dec.format(position.quantity, 0), 8)} ` +
        `${rpad(Dec.format(position.averageCost, 4), 11)} ` +
        `${rpad(Dec.format(position.marketValue), 14)} ` +
        `${rpad(Dec.format(position.unrealizedPnl), 13)}`,
    );
  }

  console.log('-'.repeat(80));
  console.log(`${pad('Total unrealized', 57)}${rpad(Dec.format(totalUnrealized), 13)}`);
  return positions;
}

// ---------------------------------------------------------------------------
// GET /accounts/{accountId}/orders
// ---------------------------------------------------------------------------

/**
 * Working orders, then a wider historical view.
 *
 * With no `status` filter the endpoint returns working orders only -- the set
 * a trading screen needs on load. Pass `status=ALL` to widen it.
 */
export async function showOrderStatus(client, account) {
  const ordersUrl = client.link(account, 'orders');

  const { items: working } = await client.get(ordersUrl);
  console.log(`\nWorking orders: ${working.length}`);

  for (const order of working) {
    const symbol = order.instrument?.symbol ?? order.strategy ?? '?';
    console.log(
      `  ${order.orderId}  ${pad(order.status, 17)} ${pad(order.side ?? '', 12)} ` +
        `${pad(symbol, 10)} ${Dec.format(order.filledQuantity, 0)}/` +
        `${Dec.format(order.quantity, 0)} @ ${order.limitPrice ?? 'MKT'}`,
    );

    // Presence of the link is the authority on what you may do next.
    const actions = ['cancel', 'replace'].filter((rel) => client.can(order, rel));
    console.log(`      available: ${actions.join(', ') || 'none (terminal)'}`);
  }

  // Everything from the last 7 days, terminal states included.
  const weekAgo = new Date(Date.now() - 7 * 864e5).toISOString();
  const recent = await client.collect(ordersUrl, { params: { status: ['ALL'], from: weekAgo } });
  console.log(`\nOrders in the last 7 days: ${recent.length}`);

  const byStatus = new Map();
  for (const order of recent) byStatus.set(order.status, (byStatus.get(order.status) ?? 0) + 1);
  for (const [status, count] of [...byStatus].sort((a, b) => b[1] - a[1])) {
    console.log(`  ${pad(status, 18)} ${count}`);
  }

  return working;
}

/**
 * Poll one order to a terminal state using conditional requests.
 *
 * `If-None-Match` makes an unchanged poll a 304 with no body, which is what
 * keeps a 1-second cadence inside the rate limit. For anything
 * latency-sensitive use the SSE stream -- this is for batch reconciliation.
 */
export async function pollOrderUntilTerminal(client, order, { timeoutMs = 60_000 } = {}) {
  const terminal = new Set(['FILLED', 'CANCELLED', 'REJECTED', 'EXPIRED', 'REPLACED']);
  const url = client.link(order, 'self');
  const deadline = Date.now() + timeoutMs;
  let etag;
  let current = order;

  while (Date.now() < deadline) {
    const { body, headers } = await client.getWithHeaders(url, {
      headers: etag ? { 'If-None-Match': etag } : {},
    });

    if (body) {
      current = body;
      etag = headers.get('ETag');
      console.log(`  -> ${current.status} filled=${current.filledQuantity}/${current.quantity}`);
      if (terminal.has(current.status)) return current;
    }

    await new Promise((r) => setTimeout(r, 1000));
  }

  console.log(`  timed out; order still ${current.status}`);
  return current;
}

/** Watch order events live instead of polling. */
export async function watchOrderEvents(client, account, { durationMs = 30_000 } = {}) {
  console.log(`\nWatching order events for ${durationMs / 1000}s...`);
  const controller = new AbortController();
  setTimeout(() => controller.abort(), durationMs);

  const seen = new Set();
  try {
    for await (const event of client.streamOrderEvents(account, { signal: controller.signal })) {
      // Delivery is at-least-once; dedupe on eventId.
      if (seen.has(event.eventId)) continue;
      seen.add(event.eventId);

      if (event.eventType === 'execution') {
        const fill = event.execution;
        console.log(
          `  fill ${event.orderId}: ${fill.quantity} @ ${fill.price} on ${fill.venue} ` +
            `-> ${event.status}`,
        );
      } else {
        console.log(`  ${event.eventType} ${event.orderId} -> ${event.status}`);
      }
    }
  } catch (err) {
    if (err.name !== 'AbortError') throw err;
  }
}

// ---------------------------------------------------------------------------
// GET /accounts/{accountId}/transactions
// ---------------------------------------------------------------------------

/**
 * Transaction history over a date window.
 *
 * Settlement can back-date entries, so a transaction may appear in a page you
 * already read. Reconcile on `transactionId`, which is permanent and unique --
 * hence the dedupe below rather than a blind sum.
 */
export async function showTransactions(client, account, { days = 30 } = {}) {
  const today = new Date();
  const params = {
    from: new Date(today.getTime() - days * 864e5).toISOString().slice(0, 10),
    to: today.toISOString().slice(0, 10),
  };

  const seen = new Set();
  const transactions = [];
  let netCash = '0';

  for await (const txn of client.paginate(client.link(account, 'transactions'), { params })) {
    if (seen.has(txn.transactionId)) continue;
    seen.add(txn.transactionId);
    transactions.push(txn);
    netCash = Dec.add(netCash, txn.netAmount);
  }

  console.log(`\nTransactions, last ${days} days: ${transactions.length}`);
  console.log(`${pad('Date', 12)} ${pad('Type', 16)} ${pad('Description', 44)} ${rpad('Net', 13)}`);
  console.log('-'.repeat(88));

  for (const txn of transactions.slice(0, 20)) {
    console.log(
      `${pad(txn.tradeDate, 12)} ${pad(txn.type, 16)} ` +
        `${pad((txn.description ?? '').slice(0, 44), 44)} ` +
        `${rpad(Dec.format(txn.netAmount), 13)}`,
    );
  }
  if (transactions.length > 20) console.log(`... and ${transactions.length - 20} more`);

  console.log('-'.repeat(88));
  console.log(`${pad('Net cash movement', 73)}${rpad(Dec.format(netCash), 13)}`);

  // Fee totals are a common reconciliation need and are not exposed as a
  // summary field, so roll them up from the fee arrays.
  const feesByType = new Map();
  for (const txn of transactions) {
    for (const fee of txn.fees ?? []) {
      feesByType.set(fee.type, Dec.add(feesByType.get(fee.type) ?? '0', fee.amount));
    }
  }
  if (feesByType.size) {
    console.log('\nFees by type:');
    for (const [type, amount] of [...feesByType].sort()) {
      console.log(`  ${pad(type, 14)} ${rpad(Dec.format(amount), 10)}`);
    }
  }

  return transactions;
}

// ---------------------------------------------------------------------------

async function main() {
  const client = new BrokerageClient();

  try {
    const accounts = await listAccounts(client);
    if (accounts.length === 0) {
      console.log('No accounts available under this grant.');
      return;
    }

    const account = accounts[0];
    console.log(`\n${'='.repeat(88)}`);
    console.log(`Account ${account.accountId} (${account.nickname ?? account.type})`);
    console.log('='.repeat(88));

    await showBalances(client, account);
    await showPositions(client, account);
    await showOrderStatus(client, account);
    await showTransactions(client, account);
  } catch (err) {
    if (!(err instanceof ApiError)) throw err;
    console.error(`\nAPI error: ${err.message}`);
    if (err.type.endsWith('/insufficient-scope')) {
      console.error(`  Re-authorize with the ${err.problem.requiredScope} scope.`);
    }
    for (const detail of err.errors) console.error(`  ${detail.pointer}: ${detail.detail}`);
    process.exitCode = 1;
  }
}

if (import.meta.url === `file://${process.argv[1]}`) await main();

File: brokerage-apis/examples/python/accounts_examples.py
"""
Accounts API examples: list, balances, positions, order status, transactions.

    python3 accounts_examples.py
"""

from __future__ import annotations

import datetime as dt
from decimal import Decimal

from brokerage_client import ApiError, BrokerageClient


# ---------------------------------------------------------------------------
# GET /accounts
# ---------------------------------------------------------------------------


def list_accounts(client: BrokerageClient) -> list[dict]:
    """Every account reachable with the current grant.

    `accountId` is the opaque handle used everywhere else in the API.
    `accountNumber` is the customer-facing value and arrives masked unless the
    grant carries `accounts:read:full` -- do not use it as a key.
    """
    response = client.get("/accounts")
    accounts = response["items"]

    for account in accounts:
        print(f"{account['accountId']}  {account['accountNumber']:>10}  "
              f"{account['type']:<7} {account['registration']:<16} "
              f"{account['status']:<10} options={account['optionsLevel']}")

        # Affordances are state-dependent: a RESTRICTED account has no
        # place-order link at all, so a UI can grey out the trade button
        # without knowing anything about compliance rules.
        if not client.can(account, "place-order"):
            print("    (not trade-enabled with this grant)")

    return accounts


# ---------------------------------------------------------------------------
# GET /accounts/{accountId}/balances
# ---------------------------------------------------------------------------


def show_balances(client: BrokerageClient, account: dict) -> dict:
    """Cash, buying power, and margin.

    Reached by following the `balances` link rather than building the URL, so
    the client keeps working if the path layout ever changes.
    """
    balances = client.follow(account, "balances")

    print(f"\nBalances as of {balances['asOf']} ({balances['currency']})")
    print(f"  Net liquidation : {Decimal(balances['netLiquidation']):>14,.2f}")
    print(f"  Cash (settled)  : {Decimal(balances['cash']['settled']):>14,.2f}")
    print(f"  Cash (total)    : {Decimal(balances['cash']['total']):>14,.2f}")
    print(f"  Withdrawable    : {Decimal(balances['cash']['withdrawable']):>14,.2f}")

    # Stock and option buying power differ materially in a margin account --
    # sizing an option order off stock buying power will get it rejected.
    buying_power = balances["buyingPower"]
    print(f"  Buying power")
    print(f"    stock         : {Decimal(buying_power['stock']):>14,.2f}")
    print(f"    option        : {Decimal(buying_power['option']):>14,.2f}")
    if "dayTrading" in buying_power:
        print(f"    day trading   : {Decimal(buying_power['dayTrading']):>14,.2f}")

    margin = balances.get("margin")
    if margin:
        print(f"  Excess liquidity: {Decimal(margin['excessLiquidity']):>14,.2f}")
        call = margin.get("marginCall")
        if call:
            # A margin call is the one balance field worth alerting on.
            print(f"  ** {call['type']}: {Decimal(call['amount']):,.2f} due {call['dueDate']} **")

    if balances.get("patternDayTrader"):
        print(f"  Day trades left : {balances['dayTradesRemaining']}")

    return balances


# ---------------------------------------------------------------------------
# GET /accounts/{accountId}/positions
# ---------------------------------------------------------------------------


def show_positions(client: BrokerageClient, account: dict) -> list[dict]:
    """Open positions with cost basis and unrealized P&L."""
    positions = list(client.paginate(client.link(account, "positions")))

    if not positions:
        print("\nNo open positions.")
        return positions

    print(f"\n{'Symbol':<24} {'Side':<6} {'Qty':>8} {'Avg cost':>11} "
          f"{'Market value':>14} {'Unrealized':>13}")
    print("-" * 80)

    total_unrealized = Decimal("0")
    for position in positions:
        instrument = position["instrument"]

        # Option positions are quantified in contracts; the display symbol is
        # the 21-character OSI string, which is unreadable in a list. Build
        # something human-facing from the option detail instead.
        if instrument["type"] == "OPTION":
            option = instrument["option"]
            label = (f"{option['underlyingSymbol']} {option['expiration'][2:]} "
                     f"{Decimal(option['strike']):g}{option['optionType'][0]}")
        else:
            label = instrument["symbol"]

        unrealized = Decimal(position["unrealizedPnl"])
        total_unrealized += unrealized

        print(f"{label:<24} {position['side']:<6} "
              f"{Decimal(position['quantity']):>8,g} "
              f"{Decimal(position['averageCost']):>11,.4f} "
              f"{Decimal(position['marketValue']):>14,.2f} "
              f"{unrealized:>+13,.2f}")

    print("-" * 80)
    print(f"{'Total unrealized':<57}{total_unrealized:>+13,.2f}")
    return positions


# ---------------------------------------------------------------------------
# GET /accounts/{accountId}/orders
# ---------------------------------------------------------------------------


def show_order_status(client: BrokerageClient, account: dict) -> list[dict]:
    """Working orders, then a wider historical view.

    With no `status` filter the endpoint returns working orders only -- the set
    a trading screen needs on load. Pass `status=ALL` to widen it.
    """
    orders_url = client.link(account, "orders")

    working = client.get(orders_url)["items"]
    print(f"\nWorking orders: {len(working)}")
    for order in working:
        filled = Decimal(order["filledQuantity"])
        total = Decimal(order["quantity"])
        symbol = order.get("instrument", {}).get("symbol", order.get("strategy", "?"))
        print(f"  {order['orderId']}  {order['status']:<17} {order.get('side', ''):<12} "
              f"{symbol:<10} {filled:g}/{total:g} @ {order.get('limitPrice', 'MKT')}")

        # Presence of the link is the authority on what you may do next.
        actions = [rel for rel in ("cancel", "replace") if client.can(order, rel)]
        print(f"      available: {', '.join(actions) or 'none (terminal)'}")

    # Everything from the last 7 days, terminal states included.
    week_ago = (dt.datetime.now(dt.timezone.utc) - dt.timedelta(days=7)).isoformat()
    recent = list(client.paginate(orders_url, params={"status": ["ALL"], "from": week_ago}))
    print(f"\nOrders in the last 7 days: {len(recent)}")

    by_status: dict[str, int] = {}
    for order in recent:
        by_status[order["status"]] = by_status.get(order["status"], 0) + 1
    for status, count in sorted(by_status.items(), key=lambda kv: -kv[1]):
        print(f"  {status:<18} {count}")

    return working


def poll_order_until_terminal(client: BrokerageClient, order: dict, timeout: int = 60) -> dict:
    """Poll one order to a terminal state using conditional requests.

    `If-None-Match` makes an unchanged poll a 304 with no body, which is what
    keeps a 1-second poll inside the rate limit. For anything latency-sensitive
    use the SSE stream instead -- this exists for batch reconciliation.
    """
    import time

    terminal = {"FILLED", "CANCELLED", "REJECTED", "EXPIRED", "REPLACED"}
    url = client.link(order, "self")
    etag = None
    deadline = time.time() + timeout

    while time.time() < deadline:
        headers = {"If-None-Match": etag} if etag else {}
        body, response_headers = client.get_with_headers(url, headers=headers)

        if body is not None:          # 200: state changed
            order = body
            etag = response_headers.get("ETag")
            print(f"  -> {order['status']} filled={order['filledQuantity']}/{order['quantity']}")
            if order["status"] in terminal:
                return order

        time.sleep(1)

    print(f"  timed out after {timeout}s; order still {order['status']}")
    return order


# ---------------------------------------------------------------------------
# GET /accounts/{accountId}/transactions
# ---------------------------------------------------------------------------


def show_transactions(client: BrokerageClient, account: dict, days: int = 30) -> list[dict]:
    """Transaction history over a date window.

    Settlement can back-date entries, so a transaction may appear in a page you
    already read. Reconcile on `transactionId`, which is permanent and unique --
    that is why the running total below dedupes rather than summing blindly.
    """
    today = dt.date.today()
    params = {"from": (today - dt.timedelta(days=days)).isoformat(), "to": today.isoformat()}

    seen: set[str] = set()
    transactions: list[dict] = []
    net_cash = Decimal("0")

    for txn in client.paginate(client.link(account, "transactions"), params=params):
        if txn["transactionId"] in seen:
            continue
        seen.add(txn["transactionId"])
        transactions.append(txn)
        net_cash += Decimal(txn["netAmount"])

    print(f"\nTransactions, last {days} days: {len(transactions)}")
    print(f"{'Date':<12} {'Type':<16} {'Description':<44} {'Net':>13}")
    print("-" * 88)
    for txn in transactions[:20]:
        print(f"{txn['tradeDate']:<12} {txn['type']:<16} "
              f"{txn.get('description', '')[:44]:<44} "
              f"{Decimal(txn['netAmount']):>+13,.2f}")
    if len(transactions) > 20:
        print(f"... and {len(transactions) - 20} more")
    print("-" * 88)
    print(f"{'Net cash movement':<73}{net_cash:>+13,.2f}")

    # Fee totals are a common reconciliation need and are not exposed as a
    # summary field, so roll them up from the fee arrays.
    fees_by_type: dict[str, Decimal] = {}
    for txn in transactions:
        for fee in txn.get("fees", []):
            fees_by_type[fee["type"]] = fees_by_type.get(fee["type"], Decimal("0")) + Decimal(fee["amount"])
    if fees_by_type:
        print("\nFees by type:")
        for fee_type, amount in sorted(fees_by_type.items()):
            print(f"  {fee_type:<14} {amount:>10,.2f}")

    return transactions


# ---------------------------------------------------------------------------


def main() -> None:
    client = BrokerageClient()

    try:
        accounts = list_accounts(client)
        if not accounts:
            print("No accounts available under this grant.")
            return

        account = accounts[0]
        print(f"\n{'=' * 88}\nAccount {account['accountId']} ({account.get('nickname') or account['type']})\n{'=' * 88}")

        show_balances(client, account)
        show_positions(client, account)
        show_order_status(client, account)
        show_transactions(client, account)

    except ApiError as exc:
        print(f"\nAPI error: {exc}")
        if exc.type.endswith("/insufficient-scope"):
            print(f"  Re-authorize with the {exc.problem.get('requiredScope')} scope.")
        for error in exc.errors:
            print(f"  {error['pointer']}: {error['detail']}")
        raise SystemExit(1)


if __name__ == "__main__":
    main()
